src/replacement/simple_replace.rs

```diff
@@ -0,0 +1,228 @@
+use std::collections::{HashMap, HashSet};
+
+use crate::{hugr::Node, Hugr, Port};
+use thiserror::Error;
+
+/// Specification of a simple replacement operation.
+#[derive(Debug, Clone)]
+pub struct SimpleReplacement {
+    /// The common DFG parent of all nodes to be replaced.
+    pub p: Node,
+    /// The set of nodes to remove (a convex set of leaf children of `p`).
+    pub s: HashSet<Node>,
+    /// A hugr with DFG root (consisting of replacement nodes).
+    pub n: Hugr,
+    /// A map from (target ports of edges from the Input node of n) to (target ports of edges from
+    /// non-s nodes to s nodes).
+    pub nu_inp: HashMap<(Node, Port), (Node, Port)>,
+    /// A map from (target ports of edges from s nodes to non-s nodes) to (input ports of the Output
+    /// node of n).
+    pub nu_out: HashMap<(Node, Port), Port>,
+}
+
+impl SimpleReplacement {
+    /// Create a new [`SimpleReplacement`] specification.
+    pub fn new(
+        p: Node,
+        s: HashSet<Node>,
+        n: Hugr,
+        nu_inp: HashMap<(Node, Port), (Node, Port)>,
+        nu_out: HashMap<(Node, Port), Port>,
+    ) -> Self {
+        Self {
+            p,
+            s,
+            n,
+            nu_inp,
+            nu_out,
+        }
+    }
+}
+
+/// Error from a [`SimpleReplacement`] operation.
+#[derive(Debug, Clone, Error, PartialEq, Eq)]
+pub enum SimpleReplacementError {
+    /// Invalid parent node.
+    #[error("Parent node is invalid.")]
+    InvalidParentNode(),
+    /// Node requested for removal is invalid.
+    #[error("A node requested for removal is invalid.")]
+    InvalidRemovedNode(),
+    /// Node in replacement graph is invalid.
+    #[error("A node in the replacement graph is invalid.")]
+    InvalidReplacementNode(),
+}
+
+#[cfg(test)]
+mod test {
+    use std::collections::{HashMap, HashSet};
+
+    use itertools::Itertools;
+    use portgraph::Direction;
+
+    use crate::builder::{BuildError, Container, Dataflow, ModuleBuilder};
+    use crate::hugr::view::HugrView;
+    use crate::hugr::{Hugr, Node};
+    use crate::ops::tag::OpTag;
+    use crate::ops::{DataflowOp, LeafOp, OpType};
+    use crate::types::{LinearType, Signature, SimpleType};
+    use crate::{type_row, Port};
+
+    use super::SimpleReplacement;
+
+    const QB: SimpleType = SimpleType::Linear(LinearType::Qubit);
+
+    /// Creates a hugr with a DFG.
```

Comment:
When we have rooted-at-any-op Hugrs, we might as well skip much of this as well, right? (Just to simplify the test)